File: sections/31_sieve_primes_factorization/problems/b_factor_signature/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<int> values(q);
    int mx = 1;
    for (int &x : values) {
        cin >> x;
        mx = max(mx, x);
    }
    vector<int> spf(mx + 1);
    iota(spf.begin(), spf.end(), 0);
    if (mx >= 1) spf[1] = 1;
    for (long long p = 2; p <= mx; p++) {
        if (spf[p] == p && p * p <= mx) {
            for (long long x = p * p; x <= mx; x += p) {
                if (spf[x] == x) spf[x] = p;
            }
        }
    }
    for (int x0 : values) {
        if (x0 == 1) {
            cout << "0 0 1\n";
            continue;
        }
        int x = x0, distinct = 0, total = 0, largest = 1;
        while (x > 1) {
            int p = spf[x];
            distinct++;
            largest = p;
            while (x % p == 0) {
                total++;
                x /= p;
            }
        }
        cout << distinct << ' ' << total << ' ' << largest << '\n';
    }
    return 0;
}

File: sections/31_sieve_primes_factorization/problems/b_factor_signature/solution.py
import sys


def spf_table(n: int) -> list[int]:
    spf = list(range(n + 1))
    if n >= 0:
        spf[0] = 0
    if n >= 1:
        spf[1] = 1
    for p in range(2, n + 1):
        if spf[p] == p and p * p <= n:
            for x in range(p * p, n + 1, p):
                if spf[x] == x:
                    spf[x] = p
    return spf


def signature(x: int, spf: list[int]) -> tuple[int, int, int]:
    if x == 1:
        return 0, 0, 1
    distinct = total = 0
    largest = 1
    while x > 1:
        p = spf[x]
        distinct += 1
        largest = p
        while x % p == 0:
            total += 1
            x //= p
    return distinct, total, largest


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    values = data[1 : 1 + q]
    spf = spf_table(max(values))
    out = []
    for x in values:
        out.append("{} {} {}".format(*signature(x, spf)))
    print("\n".join(out))


if __name__ == "__main__":
    main()
